torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,763 @@
+                        f"Unrecognized node: {node} with input node types: {input_node_types}."
+                    )
+
+        elif node.op == "output":
+            dp_strategy_map[node] = DataParallelStrategy(NodeType.NON_TENSOR, [])
+        else:
+            raise RuntimeError(f"op code {node.op} not supported")
+
+    return dp_strategy_map
+
+
+def mark_data_parallel_shardings(
+    train_step_graph: GraphModule,
+    num_parameters: int,
+    num_states: int,
+    dp_strategy_map: Dict[fx.Node, DataParallelStrategy],
+    parallel_mode: DataParallelStyle = DataParallelStyle.FULLY_SHARD,
+) -> None:
+    """
+    This function marks the sharding for the nodes in the train_step_graph
+    """
+    activation_idx = num_parameters + num_states
+    placeholder_idx = 0
+    for node in train_step_graph.graph.nodes:
+        node_type = dp_strategy_map[node].node_type
+        if node.op == "placeholder":
+            node_strategies = dp_strategy_map[node].strategies
+            if node_type == NodeType.NON_TENSOR:
+                # set node sharding to None
+                node_sharding = None
+            elif placeholder_idx < activation_idx:
+                assert len(node_strategies) > 0, "node_strategies should not be empty"
+                if parallel_mode == DataParallelStyle.REPLICATE:
+                    # set to replicate for replicate style
+                    node_sharding = node_strategies[0]
+                elif parallel_mode == DataParallelStyle.FULLY_SHARD:
+                    # set to shard for fully shard style
+                    node_sharding = node_strategies[1]
+                elif parallel_mode == DataParallelStyle.DEFAULT:
+                    # todo: add support for default mode
+                    # default mode would generate either replicate or shard
+                    raise NotImplementedError("default mode not implemented")
+            else:
+                assert len(node_strategies) > 0, "node_strategies should not be empty"
+                # mark activation as sharded on batch dim
+                node_sharding = node_strategies[0]
+
+            node.meta["sharding"] = node_sharding
+
+            placeholder_idx += 1
+        elif node.op == "call_function":
+            node_strategies = dp_strategy_map[node].strategies
+            assert (
+                len(node_strategies) <= 2
+            ), "data parallel should have at most 2 strategies"
+            if len(node_strategies) == 1:
+                node.meta["sharding"] = node_strategies[0]
+            elif len(node_strategies) == 2:
+                if parallel_mode == DataParallelStyle.REPLICATE:
+                    # set to replicate for replicate style
+                    node.meta["sharding"] = node_strategies[0]
+                elif parallel_mode == DataParallelStyle.FULLY_SHARD:
+                    # set to shard for fully shard style
+                    node.meta["sharding"] = node_strategies[1]
```

Comment:
IIUC, this means global `parallel_mode` dictates the style. I assume we plan to do smarter things later, otherwise, we don't need to generate `node_strategies` list?

If yes, let's add a TODO/FIXME